### PR TITLE
[api] Add QgsProject::crs3D

### PR DESCRIPTION
**Pull request by nyalldawson**
Returns the CRS to use for the project when transforming 3D data, or when z/elevation value handling is important.

The returned CRS will take into account verticalCrs(), e.g. by returning a compound CRS consisting of crs() + verticalCrs().

This method may still return a 2D CRS, e.g in the case that crs() is a 2D CRS and no verticalCrs() has been set for the project. I.E. It is NOT guaranteed that the returned CRS will actually be a 3D CRS, but rather it is guaranteed that the returned CRS is **ALWAYS** the most appropriate CRS to use when handling 3D data.

Refs discussion in https://github.com/qgis/QGIS/pull/56791, https://github.com/qgis/QGIS-Enhancement-Proposals/issues/267

### CODE REVIEW

**Comment by nyalldawson** [comment]
@rouault @wonder-sk as previously discussed

File: src/core/project/qgsproject.h [hunk 1]
```diff
@@ -370,13 +370,36 @@ class CORE_EXPORT QgsProject : public QObject, public QgsExpressionContextGenera
...
      */
     QgsCoordinateReferenceSystem crs() const;
 
+    /**
+     * Returns the CRS to use for the project when transforming 3D data, or when z/elevation
+     * value handling is important.
+     *
+     * The returned CRS will take into account verticalCrs(), e.g. by returning a compound
+     * CRS consisting of crs() + verticalCrs(). This method may still return a 2D CRS, e.g in the
+     * case that crs() is a 2D CRS and no verticalCrs() has been set for the project.
```
Comment (uclaros) [inline src/core/project/qgsproject.h:390]: Maybe it is worth noting how one can verify if the returned crs is actually 3d or 2d?

**Comment by wonder-sk** [comment]
@nyalldawson is this going to work if 3D Geographic (e.g. EPSG:4979) or Geocentric (e.g. EPSG:4978) CRS is set as the project CRS? Or is support for that out of scope for now?

File: src/core/project/qgsproject.h (hunk 1 above)
Comment (nyalldawson) [reply src/core/project/qgsproject.h:390]: Ok, done

**Comment by nyalldawson** [comment]
@wonder-sk I've applied some more related refinements here

File: src/core/project/qgsproject.cpp [hunk 2]
```diff
@@ -1563,6 +1597,53 @@ void QgsProject::releaseHandlesToProjectArchive()
...
+      case Qgis::CrsType::Vertical:
+        // nonsense situation
+        mCrs3D = QgsCoordinateReferenceSystem();
+        res = false;
+        break;
+
+      case Qgis::CrsType::Unknown:
+      case Qgis::CrsType::Geodetic:
+      case Qgis::CrsType::Geographic2d:
+      case Qgis::CrsType::Projected:
```
Comment (rouault) [inline src/core/project/qgsproject.cpp:1630]: some Projected CRS might already be 3D. You should probably check the axis count

**Review by rouault — APPROVED** [review]

File: src/core/project/qgsproject.cpp (hunk 2 above)
Comment (nyalldawson) [reply src/core/project/qgsproject.cpp]: @rouault 
Thanks!

Does [458b1fd](https://github.com/qgis/QGIS/pull/57107/commits/458b1fde397cabd1f268c58ad248836c0b602e5c) and https://github.com/qgis/QGIS/pull/57107/commits/6773e679c1d195cc0aeb9a4ada75fc53449af853 address this correctly?

Comment (rouault) [reply src/core/project/qgsproject.cpp]: yes, looks good

Comment (nyalldawson) [reply src/core/project/qgsproject.cpp]: Thanks @rouault !

**Pull request merged**